src/lsptoolshost/buildDiagnosticsService.ts

```diff
@@ -0,0 +1,146 @@
+            'dotnet.backgroundAnalysis.compilerDiagnosticsScope'
+        ) as AnalysisSetting;
+
+        // If compiler and analyzer diagnostics are set to "none", show everything reported by the build
+        if (analyzerDiagnosticScope === AnalysisSetting.None && compilerDiagnosticScope === AnalysisSetting.None) {
+            return diagnosticList;
+        }
+
+        // Filter the diagnostics reported by the build. Some may already be shown by live diagnostics.
+        const buildOnlyDiagnostics: vscode.Diagnostic[] = [];
+        diagnosticList.forEach((d) => {
+            if (d.code) {
+                // If it is a project system diagnostic (e.g. "Target framework out of support")
+                // then always show it. It cannot be reported by live.
+                if (this.isProjectSystemDiagnostic(d)) {
+                    buildOnlyDiagnostics.push(d);
+                }
+                // If it is a "build-only"diagnostics (i.e. it can only be found by building)
+                // then always show it. It cannot be reported by live.
+                else if (buildOnlyIds.find((b_id) => b_id === d.code)) {
+                    buildOnlyDiagnostics.push(d);
+                } else {
+                    const isAnalyzerDiagnostic = BuildDiagnosticsService.isAnalyzerDiagnostic(d);
+                    const isCompilerDiagnostic = BuildDiagnosticsService.isCompilerDiagnostic(d);
+
+                    if (
+                        (isAnalyzerDiagnostic && analyzerDiagnosticScope === AnalysisSetting.None) ||
+                        (isCompilerDiagnostic && compilerDiagnosticScope === AnalysisSetting.None)
+                    ) {
+                        // If live diagnostics are completely turned off for this type, then show the build diagnostic
+                        buildOnlyDiagnostics.push(d);
+                    } else if (isDocumentOpen) {
+                        // no-op. The document is open and live diagnostis are on. This diagnostic is already being shown.
+                    } else if (
+                        (isAnalyzerDiagnostic && analyzerDiagnosticScope === AnalysisSetting.FullSolution) ||
+                        (isCompilerDiagnostic && compilerDiagnosticScope === AnalysisSetting.FullSolution)
+                    ) {
+                        // no-op. Full solution analysis is on for this diagnostic type. All diagnostics are already being shown.
+                    } else {
+                        // The document is closed, and the analysis setting is to only show for open files.
+                        // Show the diagnostic reported by build.
+                        buildOnlyDiagnostics.push(d);
+                    }
+                }
+            }
+        });
+
+        return buildOnlyDiagnostics;
+    }
+
+    private static isCompilerDiagnostic(d: vscode.Diagnostic): boolean {
+        return d.code ? d.code.toString().startsWith('CS') : false;
```

Comment:
does `d.code?.toString()...` work?  If so, just inline these methods.

Reply:
I didn't use that because the diagnostic can be undefined, in which case I wanted to return false.

Reply:
I could add some tests for undefind diagnostics and some other edge cases.

Reply:
This probably needs a more stronger regex check of 2 letters followed by 4 digits, say regex pattern `[cC][sS][0-9]{4}`. I have seen analyzers that report diagnostics with error codes such as `CSE1000`, which would pass the current check.